lib/charms/operator_libs_linux/v0/grub.py

```diff
@@ -29,4 +37,259 @@
+# configurations into a single file like this.
+#
+# Original files:
+#  {configs}
+#
+# If you change this file, run 'update-grub' afterwards to update
+# /boot/grub/grub.cfg.
+# For full documentation of the options in this file, see:
+#   info -f grub -n 'Simple configuration'
+"""
+
+# TODO: add better description to lib
+
+
+class ValidationError(ValueError):
+    """Exception representing value validation error."""
+
+    def __init__(self, key: str, message: str) -> None:
+        self._key = key
+        self._message = message
+        super().__init__(message)
+
+    @property
+    def key(self) -> str:
+        """Get name of key, which cause this error."""
+        return self._key
+
+    @property
+    def message(self) -> str:
+        """Get error message."""
+        return self._message
+
+
+class IsContainerError(Exception):
+    """Exception if local machine is container."""
+
+
+class ApplyError(Exception):
+    """Exception if applying new config failed."""
+
+
+def _parse_config(data: str) -> Dict[str, str]:
+    """Parse config file lines.
+
+    This function is capable to update single key value if it's used with $ symbol.
+    """
+    config = {}
+    lines = data.splitlines()
+    for line in lines:
+        line = line.strip()
+        if not line or line.startswith("#"):
+            logger.debug("skipping line `%s`", line)
+            continue
+
+        key, value = line.split("=", 1)
+        if key in config:
+            logger.warning("key %s is duplicated in config", key)
+
+        config[key] = value
+
+    return config
+
+
+def _load_config(path: Path) -> Dict[str, str]:
+    """Load config file from /etc/default/grub.d/ directory."""
+    if not path.exists():
+        raise FileNotFoundError("grub config file %s was not found", path)
+
+    with open(path, "r", encoding="UTF-8") as file:
+        data = file.read()
+        config = _parse_config(data)
+
+    logger.info("grub config file %s was loaded", path)
+    logger.debug("config file %s", config)
+    return config
+
+
+def _save_config(path: Path, config: Dict[str, str], header: str = CONFIG_HEADER) -> None:
+    """Save grub config file."""
+    if path.exists():
+        logger.debug("grub config %s already exist and it will overwritten", path)
+
+    context = [f"{key}={value}" for key, value in config.items()]
+    with open(path, "w", encoding="UTF-8") as file:
+        file.writelines([header, *context])
+
+    logger.info("grub config file %s was saved", path)
+
+
+def is_container() -> bool:
+    """Help function to see if local machine is container."""
+    # TODO: implement it
+    return False
+
+
+class GrubConfig(Mapping[str, str]):
+    """Grub config object.
+
+    This object will load current configuration option for grub and provide option
+    to update it with simple validation.
+
+    There is only one public function `update`, which should handle everything. It will
```

Comment:
There's also `remove`, by the looks. Might be better not to mention the public methods here to avoid this getting out of date. Perhaps define the public methods at the top (after `__init__`) so they're more discoverable?

Reply:
You are ring, I forgot that I add `remove` and `apply`. I'll update docstring.
I'm used to define all public on bottom (my order magic objects (__init__, __eq__, ...), private properties, private functions, public properties and public method), but I'm ok to move it to top if that required.